### DISCUSSION TITLE
Commit V.S. DeliverTx

### DISCUSSION THREAD

**Discussion by pcpLiu** [tendermint/spec#381] [Q&A] [answered]
Hi team,

The Commit [doc](https://docs.tendermint.com/v0.35/spec/abci/abci.html#commit) says

```
Commit signals the application to persist application state. It takes no parameters.
```

For DeliverTx, [the doc](https://docs.tendermint.com/v0.35/spec/abci/abci.html#delivertx) says:

```
...The transaction was already included in a block, so the Code does not influence Tendermint consensus.
```

I have two understandings:
- a) Whenever our application receives a `DeliverTx` call, we can write current app state after this TX into our persistent storage as an immutable truth. And when we receive the `Commit` call, we just know that this block is finalized. We can do something. 
- b) All `DeliverTx` calls before `Commit` are not finalized. We should only write the app state into local DB when we receive `Commit` call.

I think a) is the right understanding but wanna double check with you.

Thanks.

**Comment by creachadair** [comment, marked as answer]
Hi @pcpLiu. In general, it isn't safe for the application to update its state until a block is fully committed. The application can execute and stage delivered transactions to persistent storage if it wants, but the effects of those transactions should not be exposed to users of the application until after a successful commit.

Otherwise, suppose the consensus node or the application itself fail during delivery: After restart, a node will replay the block(s) from the last-known-good application height. Unless the transactions are idempotent, replaying the same transactions may change the application's state differently than the original transactions would have done.

So the safe answer is closer to (b).

**Reply by pcpLiu** [reply]
Thanks! The explanation is making a lot of sense.
